### PR TITLE
BUGFIX | Fix `autofix` misplacing `}` when resolving unused function parameters, and captures.

### PR DESCRIPTION
**Pull request by alexatcanva**
# What is this ? 

Hi All :wave: 

This is my first PR contributing to `zls`, and it's also my first Zig PR (ever), so if at any point during this PR process I'm a little slow to pick things up please forgive me :pray:

This PR is aimed at resolving issue https://github.com/zigtools/zls/issues/1543 

The issue occurs when you use the `autofix` functionality of `zls` to automatically provide discarded variable usages inside your function block, or inside your capture block. 

An example can be seen like so: 

```zig 

pub fn format(a: u8, b: u8, c: u8) !void {
}

// Then after you save:


pub fn format(a: u8, b: u8, c: u8) !void {
    _ = c;
    _ = b;
    _ = a;}

```

This PR addresses this issue so the output of the `autofix` should now be like: 

```zig

pub fn format(a: u8, b: u8, c: u8) !void {
    _ = a;
    _ = b;
    _ = c;
}
```

This behaviour can also be noted for the capture discards: 

```zig

// old

pub fn format(a: u8, b: u8, c: u8) !void {
    _ = c;
    _ = b;
    _ = a;

    for (0..10) |ix| {
        _ = ix;}
}

// new

pub fn format(a: u8, b: u8, c: u8) !void {
    _ = a;
    _ = b;
    _ = c;

    for (0..10) |ix| {
        _ = ix;
    }
}

```

Let me know if I've missed anything here! Thanks!

### CODE REVIEW

File: src/features/code_actions.zig [hunk 1]
```diff
@@ -509,6 +558,16 @@ fn getCaptureLoc(text: []const u8, loc: offsets.Loc) ?offsets.Loc {
...
 }
 
+fn prependSlice(list: *std.ArrayListUnmanaged(types.CodeAction), arena: std.mem.Allocator, values: []const types.CodeAction) !void {
+    const old_len = list.items.len;
+    try list.resize(arena, list.items.len + values.len);
+    var i = old_len + values.len - 1;
+    while (i >= values.len) : (i -= 1) {
+        list.items[i] = list.items[i - values.len];
+    }
+    std.mem.copy(types.CodeAction, list.items[0..values.len], values);
```
Comment (Techatrix) [inline src/features/code_actions.zig:568]: The `insertSlice` function should provide what you need. I would suggest to remove the `prependSlice` function all together and just call `insertSlice` directly.
```zig
try list.insertSlice(arena, 0, values);
```

File: src/features/code_actions.zig [hunk 2]
```diff
@@ -109,7 +109,35 @@ fn handleUnusedFunctionParameter(builder: *Builder, actions: *std.ArrayListUnman
...
+    // To do this, you may have a function like:
+    // fn(a: i32, b: i32, c: i32) void { _ = a; _ = b; _ = c; }
+    // or
+    // fn(
+    //     a: i32,
+    //     b: i32,
+    //     c: i32,
+    // ) void { ... }
+    // We have to be able to detect both cases.
+    var param_end = offsets.tokenToLoc(tree, ast.paramLastToken(tree, payload.get(tree).?)).end;
```
Comment (Techatrix) [inline src/features/code_actions.zig:123]: Further towards the end of this function `payload.get(tree)` is being called again. I would suggest to do it once and store into a new variable.

File: src/features/code_actions.zig [hunk 3]
```diff
@@ -109,7 +109,35 @@ fn handleUnusedFunctionParameter(builder: *Builder, actions: *std.ArrayListUnman
...
+                found_comma = true;
+                continue;
+            } else {
+                param_end += 1;
+                break;
+            },
+            ')' => break,
+            else => break,
+        }
+    }
```
Comment (Techatrix) [inline src/features/code_actions.zig:138]: Looks like you copied this code from the `getParamRemovalRange` function. You do not use `param_end` after iterating so there is no need to continue after encountering a comma.

This means that we can write the following:
```zig
var found_comma: bool = false;
while (param_end < tree.source.len) : (param_end += 1) {
    switch (tree.source[param_end]) {
        ' ', '\n' => continue,
        ',' => {
            found_comma = true;
            break;
        },
        ')' => break,
        else => break,
    }
}
```

If we use some of the functions from the `std.mem` we can also do this:
```zig
const found_comma = std.mem.startsWith(u8, std.mem.trimLeft(u8, tree.source[param_end..], " \n"), ",");
```

File: src/features/code_actions.zig [hunk 4]
```diff
@@ -310,7 +352,7 @@ fn createCamelcaseText(allocator: std.mem.Allocator, identifier: []const u8) ![]
 }
 
 // returns a discard string `\n{indent}_ = identifier_name;`
-fn createDiscardText(builder: *Builder, identifier_name: []const u8, declaration_start: usize, add_block_indentation: bool) ![]const u8 {
+fn createDiscardText(builder: *Builder, identifier_name: []const u8, declaration_start: usize, add_block_indentation: bool, add_suffix_newline: bool) ![]const u8 {
```
Comment (Techatrix) [inline src/features/code_actions.zig:355]: Could you add a comma after the last parameter and then run `zig fmt` so the parameters gets spread out over multiple lines?

File: src/features/code_actions.zig [hunk 5]
```diff
@@ -321,16 +363,23 @@ fn createDiscardText(builder: *Builder, identifier_name: []const u8, declaration
...
 
     new_text.appendAssumeCapacity('\n');
     new_text.appendSliceAssumeCapacity(indent);
     new_text.appendSliceAssumeCapacity(additional_indent);
     new_text.appendSliceAssumeCapacity("_ = ");
     new_text.appendSliceAssumeCapacity(identifier_name);
     new_text.appendAssumeCapacity(';');
+    // if we are adding a suffix newline, we need to add the additional indentation again.
+    new_text.appendSliceAssumeCapacity(additional_newline);
+    if (add_suffix_newline) new_text.appendSliceAssumeCapacity(indent);
```
Comment (Techatrix) [inline src/features/code_actions.zig:382]:
```suggestion
    if (add_suffix_newline) {
        new_text.appendAssumeCapacity('\n');
        new_text.appendSliceAssumeCapacity(indent);
    }
```

File: src/features/code_actions.zig [hunk 6]
```diff
@@ -321,16 +363,23 @@ fn createDiscardText(builder: *Builder, identifier_name: []const u8, declaration
...
     const additional_indent = if (add_block_indentation) detectIndentation(builder.handle.tree.source) else "";
+    const additional_newline = if (add_suffix_newline) "\n" else "";
 
-    const new_text_len = 1 + indent.len + additional_indent.len + "_ = ;".len + identifier_name.len;
-    var new_text = try std.ArrayListUnmanaged(u8).initCapacity(builder.arena, new_text_len);
+    var new_text_len = 1 + indent.len + additional_indent.len + additional_newline.len + "_ = ;".len + identifier_name.len;
+    // if we have an additional newline, then we will need an additional indent too.
+    if (add_suffix_newline) new_text_len += indent.len;
+    const post_new_len = new_text_len;
+    var new_text = try std.ArrayListUnmanaged(u8).initCapacity(builder.arena, post_new_len);
```
Comment (Techatrix) [inline src/features/code_actions.zig:372]:
```suggestion
    const new_text_len =
        1 +
        indent.len +
        additional_indent.len +
        "_ = ;".len +
        identifier_name.len +
        if (add_suffix_newline) 1 + indent.len else 0;
    var new_text = try std.ArrayListUnmanaged(u8).initCapacity(builder.arena, new_text_len);
```

**Review by Techatrix — CHANGES_REQUESTED** [review]
Hi @alexatcanva  👋

I've had some comment but those are mostly code simplification or formatting and nothing critical. Overall you did a good job.

The most important thing is the missing test coverage. The current code action test were formatting the source code before comparing so there was no way to check for correctly formatted output.

The following patch should remove the formatting pass (and adjust one test case).
Most code action tests fail after this patch but not after your PR which shows that you produce code that is formatted just like `zig fmt`.

```patch
diff --git a/tests/lsp_features/code_actions.zig b/tests/lsp_features/code_actions.zig
index 5f9ae04..8d9796c 100644
--- a/tests/lsp_features/code_actions.zig
+++ b/tests/lsp_features/code_actions.zig
@@ -85,10 +85,11 @@ test "code actions - discard captures" {
 test "code actions - discard capture with comment" {
     try testAutofix(
         \\test {
-        \\  if (1 == 1) |a|
-        \\      //a
-        \\      {}
+        \\    if (1 == 1) |a|
+        \\    //a
+        \\    {}
         \\}
+        \\
     ,
         \\test {
         \\    if (1 == 1) |a|
@@ -174,9 +175,5 @@ fn testAutofix(before: []const u8, after: []const u8) !void {
     defer allocator.free(actual);
     try ctx.server.document_store.refreshDocument(uri, try allocator.dupeZ(u8, actual));
 
-    try std.testing.expect(handle.tree.errors.len == 0);
-    const formatted_actual = try handle.tree.render(allocator);
-    defer allocator.free(formatted_actual);
-
-    try std.testing.expectEqualStrings(after, formatted_actual);
+    try std.testing.expectEqualStrings(after, handle.tree.source);
 }
```

**Comment by alexatcanva** [comment]
Wow! @Techatrix thank you so much for taking the time to provide such an in-depth review! 🙏 

I really appreciate the effort you have spent here to clearly outline and explain the improvements that I could have made in here. This type of feedback is invaluable while I am trying to learn the ins-and-outs of Zig and zls! 

I have made the relevant adjustments to this PR as per your review comments, please let me know if there's anything else I can do to assist further! 

Thanks again! 🚀

**Review by Techatrix — APPROVED** [review]
:+1:

**Pull request merged**